### PR TITLE
[check] call builder

### PR DESCRIPTION
**Pull request by alangenfeld**
Adds a utility funtion to check for programatiically building calls based on a type signatures. 

## How I Tested These Changes

added tests

### CODE REVIEW

**Comment by alangenfeld** [comment]
* **#22043** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22043?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22260** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22260?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @alangenfeld and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_check/__init__.py [hunk 1]
```diff
@@ -1805,3 +1832,201 @@ def _check_two_dim_mapping_entries(
...
+                return ref._evaluate(self.global_ns, self.local_ns)  # noqa
+            else:
+                return ref._evaluate(self.global_ns, self.local_ns, frozenset())  # noqa
+        except NameError as e:
+            raise CheckError(f"Unable to resolve {ref}") from e
+
+    def update_from_frame(self, depth: int):
+        ctx_frame = sys._getframe(depth + 1)  # noqa # surprisingly not costly
+        self.global_ns.update(ctx_frame.f_globals)
+        self.local_ns.update(ctx_frame.f_locals)
```
Comment (schrockn) [inline python_modules/dagster/dagster/_check/__init__.py:1864]: Could you add a docblock to describe what is going on here? I think I am inferring quite a bit but docblock would be very helpful.

**Review by schrockn — CHANGES_REQUESTED** [review]
Looks fantastic overall.

How hard would this be to apply to arbitrary function calls?

e.g.

```python
@checked
def foo(bar: int, baz: str) -> None:
  ...
```

Also do we have a sense of perf?

**Review by schrockn — COMMENTED** [review]
Also can you add test cases demonstrating the features of `EvalContext`?

Also demonstration of some error messages and stack traces in test plan would be nice too.

File: python_modules/dagster/dagster/_check/__init__.py [hunk 2]
```diff
@@ -1805,3 +1832,201 @@ def _check_two_dim_mapping_entries(
...
+                    if inner_origin is list:
+                        return partial(
+                            opt_nullable_list_param, param_name=name, of_type=inner_single
+                        )
+                    elif inner_origin is dict:
+                        return partial(
+                            opt_nullable_dict_param,
+                            param_name=name,
+                            key_type=inner_pair_left,
+                            value_type=inner_pair_right,
```
Comment (schrockn) [inline python_modules/dagster/dagster/_check/__init__.py:2010]: avoiding indenttation this deep would be nice

**Comment by alangenfeld** [comment]
> How hard would this be to apply to arbitrary function calls?

pretty sure trivial , was planning on adding in follow up

**Comment by alangenfeld** [comment]
> Also do we have a sense of perf?

This is more about the top of the stack, but:

before and after using `scripts/object_profile.py` for converting `AssetKey` https://gist.github.com/alangenfeld/2f43d84748ed5dfd99849c665af18d82
no memory increase but time goes from ~334 to ~730 ns per call

and for this shoot out script: https://gist.github.com/alangenfeld/23c3e62fd1afe52136b73f4d533c0536
these results: https://gist.github.com/alangenfeld/ea2505b7923642bb78be852d8f5ebb35

so in this iteration, a little faster than pydantic but still slower than manual.

Banging on it a bit locally it seems to just be the extra function call overhead. I think using `exec` to generate the same code as what would be in the manual call is the next thing to explore.

**Review by schrockn — DISMISSED** [review]
Req'ing based on docs for `EvalContext` and optionally reducing indentation of the workhorse fn

**Review by schrockn — APPROVED** [review]
[make-it-so-picard](https://app.graphite.dev//settings/meme-library?org=dagster-io)

**Pull request merged**